File: src/atcoder/abc310/abc310_c.scala
import java.util.Scanner
import scala.collection.mutable.Set

object Main extends App {
  val sc = new Scanner(System.in)
  val n = sc.nextInt

  val uniq = Set[String]()
  val count = Set[String]()

  for (i <- 0 until n) {
    val s = sc.next
    val r = s.reverse

    if (s == r) {
      uniq += s
    } else {
      count += s
      count += r
    }
  }

  val ans = uniq.size + count.size / 2
  println(ans)
  
}

File: src/atcoder/abc310/abc310_c.swift
let n = Int(readLine()!)!
var uniq = Set<String>()
var count = Set<String>()

for _ in 0..<n {
  let s = readLine()!
  let r = String(s.reversed())
  if s == r {
    uniq.insert(s)
  } else {
    count.insert(s)
    count.insert(r)
  }
}

print(count.count / 2 + uniq.count)